R/slide.R

```diff
@@ -225,6 +248,67 @@ epi_slide <- function(x, f, ..., before, after, ref_time_values,
     after <- 0L
   }
 
+  if (autocomplete_windows) {
+    assert_list(fill, names = "unique")
+    fill_vars_not_in_x <- !(names(fill) %in% colnames(x))
+    if (any(fill_vars_not_in_x)) {
+      Warn(
+        c(
+          "Some names provided in `fill` do not correspond to column names in
+           the input data. Any columns not specified in `fill` will be
+           completed with missing values (`NA`) by default."
+        ),
+        class = "epiprocess__epi_slide__fill_vars_not_in_x",
+        epiprocess__fill = fill,
+        epiprocess__colnames_x = colnames(x),
+        epiprocess__fill_vars_not_in_x = fill_vars_not_in_x
+      )
+    }
+
+    key_cols <- key_colnames(x)
+    maybe_first_duplicate_key_row_index <- anyDuplicated(x, by = key_cols)
+    if (maybe_first_duplicate_key_row_index != 0L) {
+      Abort("`x` must have one row per unique combination of the key variables to use window completion. If you have additional key variables other than `geo_value`, `time_value`, and `version`, such as an age group column, please specify them in `other_keys`.  Otherwise, check for duplicate rows and/or conflicting values for the same measurement.",
+        class = "epiprocess__epi_slide__epi_df_requires_unique_key_for_completion"
+      )
+    }
+
+    # Make a complete date sequence between min(x$time_value) and max
+    # (x$time_value), plus pad values. We need to include pad dates in
+    # the complete date sequence so that the first and last n - 1
+    # ref_time_values have a complete window to use.
+    date_seq_list <- full_date_seq(x, before, after, time_step)
+    all_dates <- c(
+      date_seq_list$pad_early_dates,
+      date_seq_list$all_dates,
+      date_seq_list$pad_late_dates
+    )
+
+    # A helper column marking real observations.
+    x$.real <- TRUE
+    # Fill the `.real` column with `FALSE` for rows added during completion.
+    fill$.real <- FALSE
+
+    key_cols_no_time <- kill_time_value(key_cols)
+
+    # Add in rows for each present key column combination for all dates.
+    #  - This happens within each group if the data is grouped. This
+    #    doesn't change the impact of completion because we apply the
+    #    computation by group as well.
+    #  - If a geo first appears halfway through the dataset, it will be
+    #    completed all the way back to the beginning of the data.
+    x <- tidyr::complete(x,
+        expand(x, nesting(!!key_cols_no_time), data.frame(time_value = all_dates)),
```

Comment:
Ouch, this is complicated.  Two issues.  They feel a bit edge-case-y, but matter more if we're trying to have some consistency with `epix_slide`, or maybe a future `epi_slide_reframe` that doesn't try to match the existing keys in the output.
- If reporting starts for one epikey later on, or if there are large gaps in its reporting, then even without data revisions, you could have later epi_slides on expanded epi_dfs outputting different results for earlier ref_time_values, as they will be considering additional epikeys.  If grouping by all epikeys, then maybe this can't happen for `epi_slide` (but could for an analogous `epix_slide`).  But if you don't group by anything, you might turn earlier national sums into a bunch of NAs.
- Should we be using only the epikeys that have appeared, or the cross product of all seen geos x all seen age groups x etc.?  If we have to pick one, I think the current approach of using only epikeys that have appeared is right; user can `complete` or join or whatever if they don't like that.  (Again, for `epi_slide`, this probably only matters about whether we make not-grouped-by-all-epikeys epi_slides output NAs...)

Reply:
One more thing: I couldn't find where the matching-old-epikey-time_value/broadcasting part is.  I wonder if it's using this completed `x`.  If it is, then the two questions above actually are more relevant for `epi_slide` since it involves also changing the epikeysets output for earlier time values simply by adding an epikey in rows for later time_values, and would also mean outputting values for gaps, and maybe also NA's at the leading and trailing edges.  Ideally we should consciously make decisions here, or make it configurable.

Reply:
(you may be able to get rid of the `expand` call)

Reply:
A full_seq may be missing in definition of all_dates.  If there is a gap for all epikeys then completion may not fill it in.

Reply:
`full_date_seq` fn below should take care of that. The three cases ([tsibble class or numeric](https://github.com/cmu-delphi/epiprocess/blob/adbdcd414cf8d1366af2bb08eded088dc6eb51b8/R/slide.R#L797), [else `time_step` not defined](https://github.com/cmu-delphi/epiprocess/blob/adbdcd414cf8d1366af2bb08eded088dc6eb51b8/R/slide.R#L833), [else `time_step` defined](https://github.com/cmu-delphi/epiprocess/blob/adbdcd414cf8d1366af2bb08eded088dc6eb51b8/R/slide.R#L852)) all generate date sequences by stepping by `time_step` (or similar) between the input data's min and max time_values.

Reply:
Key cols are broadcast backwards here:
```
expand(x, nesting(!!key_cols_no_time), data.frame(time_value = all_dates))
```
This wouldn't change output epikeysets for earlier time values. In the non-aggregating case, computations are both completed across and computed across all keycols, and we later (at the end of this function) filter out not-`.real` obs. Epikeysets that were broadcast back in time will be removed by the filter.

In the aggregating case (e.g. states -> national sum), we're outputting all-new epikeysets. We should end up with one row per group. As long as the input group data contains one `.real = TRUE` obs, the output should be kept (?)(I'm not sure how `.real` is handled right now -- need to check/implement that logic).